### PR TITLE
Pagination in cloud projects

### PR DESCRIPTION
**Pull request by mohsenD98**
This PR aims to improve the user experience and reduce waiting time by fetching the cloud projects list page by page, allowing the UI to load faster and more efficiently. This optimization enables users to quickly see the project list without having to wait for large amounts of data to be loaded at once, resulting in a smoother and more responsive user experience.

###### remaining tasks: 
- [x] use `insertRow` instead of `resetModel`.   
- [x] set offset and limit in a QNetworkRequest.
- [x] remove debug messages.  

**Demo:**

[Screencast from 2024-08-13 20-21-33.webm](https://github.com/user-attachments/assets/a925b093-b772-4c68-a99a-52ce0fe590f7)

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 439432409835ab0631f7da2abde7cf0d6eb246a3: [**arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5558-2450f306bbcd977b5a740b972689d3d610dfced8-android-arm64.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5558-2450f306bbcd977b5a740b972689d3d610dfced8-android-arm-neon.apk)
- [x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5558-2450f306bbcd977b5a740b972689d3d610dfced8-android-x64.apk)
- [x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5558-2450f306bbcd977b5a740b972689d3d610dfced8-android-x86.apk)
- [**all access arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5558-2450f306bbcd977b5a740b972689d3d610dfced8-android-arm64.apk)
- [all access arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5558-2450f306bbcd977b5a740b972689d3d610dfced8-android-arm-neon.apk)
- [all access x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5558-2450f306bbcd977b5a740b972689d3d610dfced8-android-x64.apk)
- [all access x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5558-2450f306bbcd977b5a740b972689d3d610dfced8-android-x86.apk)
</details>

File: src/core/qfieldcloudprojectsmodel.cpp [hunk 1]
```diff
@@ -1689,6 +1703,16 @@ void QFieldCloudProjectsModel::projectListReceived()
   QJsonDocument doc = QJsonDocument::fromJson( response );
   QJsonArray projects = doc.array();
   reload( projects );
+
+  if ( projects.size() > 0 )
+  {
+    qDebug() << "Fetch list size :" << projects.size();
+    refreshProjectsList( refreshPublicProjects );
```
Comment (nirvn) [inline src/core/qfieldcloudprojectsmodel.cpp:1710]: Here, use rawReply->attribute( QNetworkRequest::User + {1, 2} ) to fetch offset and whether we fetch public projects or not, then use those attributes for the refreshProjectList parameters:

```suggestion
    refreshProjectsList( isPublic, offset );
```

File: src/core/qfieldcloudprojectsmodel.cpp [hunk 2]
```diff
@@ -1689,6 +1703,16 @@ void QFieldCloudProjectsModel::projectListReceived()
...
+  }
+  else
+  {
+    qDebug() << "=============Finish==============";
```
Comment (nirvn) [inline src/core/qfieldcloudprojectsmodel.cpp:1714]: Let's get rid of the debugs throughout.

File: src/core/qfieldcloudprojectsmodel.h [hunk 3]
```diff
@@ -487,6 +491,8 @@ class QFieldCloudProjectsModel : public QAbstractListModel
     QString getJobTypeAsString( JobType jobType ) const;
 
     void downloadFileConnections( const QString &projectId, const QString &fileName );
+    void insertProjectList( const QList<CloudProject *> &projects );
+    bool isProjectExists( CloudProject *project );
```
Comment (nirvn) [inline src/core/qfieldcloudprojectsmodel.h:495]: I don't think you'll need that here, just make sure that the insertProjects function checks for duplicates.

File: src/core/qfieldcloudprojectsmodel.h [hunk 4]
```diff
@@ -487,6 +491,8 @@ class QFieldCloudProjectsModel : public QAbstractListModel
     QString getJobTypeAsString( JobType jobType ) const;
 
     void downloadFileConnections( const QString &projectId, const QString &fileName );
+    void insertProjectList( const QList<CloudProject *> &projects );
```
Comment (nirvn) [inline src/core/qfieldcloudprojectsmodel.h:494]:
```suggestion
    void insertProjects const QList<CloudProject *> &projects );
```

File: src/qml/QFieldCloudScreen.qml [hunk 5]
```diff
@@ -594,12 +595,12 @@ Page {
     }
   }
 
-  function refreshProjectsList(shouldRefreshPublic) {
+  function refreshProjectsList(shouldRefreshPublic, resetOffset) {
```
Comment (nirvn) [inline src/qml/QFieldCloudScreen.qml:598]: We're getting rid of the offset reset parameter, you'll just need to pass on the one public vs private boolean like before).

File: src/qml/QFieldCloudScreen.qml [hunk 6]
```diff
@@ -594,12 +595,12 @@ Page {
...
     if (cloudConnection.state !== QFieldCloudConnection.Idle && cloudProjectsModel.busyProjectIds.length === 0) {
       return;
     }
     table.refreshing = true;
-    cloudProjectsModel.refreshProjectsList(shouldRefreshPublic);
+    cloudProjectsModel.refreshProjectsList(shouldRefreshPublic, resetOffset);
```
Comment (nirvn) [inline src/qml/QFieldCloudScreen.qml:603]: We're getting rid of the offset reset parameter, you'll just need to pass on the one public vs private boolean like before).

**Comment by nirvn** [comment]
@mohsenD98 , here's a basic attribute example here: https://pastebin.com/XTmA5nha

**Comment by mohsenD98** [comment]
@nirvn, thank you for you grate feedback, :+1:  also we need to decide about pagination fetch per request. currently its 500 projects per fetch.

File: src/core/qfieldcloudprojectsmodel.cpp (hunk 2 above)
Comment (mohsenD98) [reply src/core/qfieldcloudprojectsmodel.cpp]: Done.

File: src/core/qfieldcloudprojectsmodel.cpp [hunk 7]
```diff
@@ -1644,6 +1657,7 @@ void QFieldCloudProjectsModel::projectCancelUpload( const QString &projectId )
 void QFieldCloudProjectsModel::connectionStatusChanged()
 {
   refreshProjectsList();
+  refreshPublicProjects = false;
```
[the remark this answers is not in the document]
Comment (mohsenD98) [reply src/core/qfieldcloudprojectsmodel.cpp]: Done.

File: src/core/qfieldcloudprojectsmodel.cpp [hunk 8]
```diff
@@ -222,14 +222,27 @@ QVariantMap QFieldCloudProjectsModel::getProjectData( const QString &projectId )
...
       QString url = shouldRefreshPublic ? QStringLiteral( "/api/v1/projects/public/" ) : QStringLiteral( "/api/v1/projects/" );
-      NetworkReply *reply = mCloudConnection->get( url );
+
+      QVariantMap params;
+      params["limit"] = QString::number( limit );
+      params["offset"] = QString::number( offset );
+
+      offset += limit;
+
+      NetworkReply *reply = mCloudConnection->get( url, params );
```
[the remark this answers is not in the document]
Comment (mohsenD98) [reply src/core/qfieldcloudprojectsmodel.cpp]: Done :+1:

File: src/core/qfieldcloudprojectsmodel.h [hunk 9]
```diff
@@ -317,6 +317,10 @@ class QFieldCloudProjectsModel : public QAbstractListModel
     void layerObserverLayerEdited( const QString &layerId );
 
   private:
+    int offset = 0;
+    int limit = 500;
+    bool refreshPublicProjects = false;
```
[the remark this answers is not in the document]
Comment (mohsenD98) [reply src/core/qfieldcloudprojectsmodel.h]: Grate notes, thanks. Done :+1:

File: src/core/qfieldcloudprojectsmodel.cpp [hunk 10]
```diff
@@ -222,15 +222,29 @@ QVariantMap QFieldCloudProjectsModel::getProjectData( const QString &projectId )
   return data;
 }
 
-void QFieldCloudProjectsModel::refreshProjectsList( bool shouldRefreshPublic )
+void QFieldCloudProjectsModel::refreshProjectsList( bool shouldRefreshPublic, int projectFetchOffset )
```
Comment (nirvn) [inline src/core/qfieldcloudprojectsmodel.cpp:225]: When projectFetchOffset is 0, reset model (so a project refresh removes projects that have disappeared on the cloud through a/ project deletion, b/ shared project revocation, c/ etc. :)

File: src/core/qfieldcloudprojectsmodel.h [hunk 11]
```diff
@@ -174,6 +174,13 @@ class QFieldCloudProjectsModel : public QAbstractListModel
       DeltaUploaded
     };
 
+    //! whether to fetch public projects and to specify the offset for pagination.
```
Comment (nirvn) [inline src/core/qfieldcloudprojectsmodel.h:177]:
```suggestion
    //! Attributes controlling fetching of projects
```

**Review by nirvn — COMMENTED** [review]
It's working really nicely, nice to see feedback within seconds instead of minutes!

File: src/core/qfieldcloudprojectsmodel.cpp (hunk 10 above)
Comment (mohsenD98) [reply src/core/qfieldcloudprojectsmodel.cpp:225]: Done.

File: src/core/qfieldcloudprojectsmodel.h [hunk 12]
```diff
@@ -174,6 +174,13 @@ class QFieldCloudProjectsModel : public QAbstractListModel
       DeltaUploaded
     };
 
+    //! Attributes controlling fetching of projects
+    enum class ProjectsRequestAttribute
```
[the remark this answers is not in the document]
Comment (m-kuhn) [reply src/core/qfieldcloudprojectsmodel.h:178]: @nirvn @mohsenD98 I would vote to silence these warnings?

[the remark this answers is not in the document]
Comment (nirvn) [reply src/core/qfieldcloudprojectsmodel.h:178]: @m-kuhn , I put you to it ;-P

[the remark this answers is not in the document]
Comment (mohsenD98) [reply src/core/qfieldcloudprojectsmodel.h:178]: Yes :+1:

**Review by nirvn — APPROVED** [review]

**Pull request merged**